### PR TITLE
Refactor bhp from thp limit

### PR DESCRIPTION
**Pull request by totto82**
This is the first of series of PR that aim at reducing the code duplication for bhpFromThpLimit between STW, MSW, producers and injectors. 

From MSW this is a pure refactoring. From STW it is not since it starts to use the same code as MSW. 

The implementation that has been used for MSW is faster and also sufficiently robust especially thanks to @GitPaean latest additions. I have tested this on several models both simple and full-field models with STW and the results looks promising. 

I will continue to test this on the norne prediction cases.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
```
mpi.compareECLFiles_flow+UDQ_M2 ok
mpi.compareECLFiles_flow+UDQ_M1 ok
mpi.compareECLFiles_flow+UDQ_M3 ok
mpi.compareECLFiles_flow+9_2A_DEPL_GCONPROD_1L_STW (31 != 30)
mpi.compareECLFiles_flow+9_1B_DEPL_MAX_RATE_MIN_THP_STW ok
mpi.compareECLFiles_flow+9_2B_DEPL_GCONPROD_2L_STW ok!
mpi.compareECLFiles_flow+9_3A_GINJ_REIN-G_STW (30 != 31)
mpi.compareECLFiles_flow+9_3B_GINJ_GAS_EXPORT_STW (30 != 31)
mpi.compareECLFiles_flow+9_3C_GINJ_GAS_GCONSUMP_STW ok-
mpi.compareECLFiles_flow+9_3D_GINJ_GAS_MAX_EXPORT_STW (32 != 31)
mpi.compareECLFiles_flow+9_3E_GAS_MIN_EXPORT_STW ok
mpi.compareECLFiles_flow+9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_STW (32 != 31)
mpi.compareECLFiles_flow+9_4B_WINJ_VREP-W_STW (31 != 32)
mpi.compareECLFiles_flow+9_4C_WINJ_GINJ_VREP-W_REIN-G_STW (32 != 31)
mpi.compareECLFiles_flow+9_4D_WINJ_GINJ_GAS_EXPORT_STW (33 != 31)
mpi.compareECLFiles_flow+ACTIONX_GCONPROD ok
mpi.compareECLFiles_flow+ACTIONX_WCONINJE (100 != 101
mpi.compareECLFiles_flow+MOD4_UDQ_ACTIONX (91 != 92)
mpi.compareECLFiles_flow+MOD4_GRP (95 != 99)
mpi.compareECLFiles_flow+MOD4_GRP_GEFAC (97 != 94)
mpi.compareECLFiles_flow+MIN_THP_1 (58 != 50) ok!
```

**Comment by totto82** [comment]
The tests are fine.

**Comment by totto82** [comment]
> The tests are fine.

There is a significant run-time increase for some of the cases that needs to be investigated.

**Comment by blattms** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 0.991    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1.002    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 0.998    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 1.013    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 1        |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 0.994    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1519

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
`There is a significant run-time increase for some of the cases that needs to be investigated.`

The issue is fixed. Now only cases with STW are affected as expected.

**Comment by GitPaean** [comment]
I am looking at this PR now. Will give my comments soon. Sorry for the delay.

File: opm/simulators/wells/StandardWellGeneric.cpp [hunk 1]
```diff
@@ -199,192 +199,10 @@ StandardWellGeneric<Scalar>::
...
-    for (int ii = 0; ii < num_samples; ++ii) {
-        dbgmsg += "  " + std::to_string(fbhp_samples[ii]);
-    }
-    OpmLog::debug(dbgmsg);
-#endif // EXTRA_THP_DEBUGGING
-
-    // Look for sign changes for the (fbhp_samples - bhp_samples) piecewise linear curve.
-    // We only look at the valid
-    int sign_change_index = -1;
-    for (int ii = 0; ii < num_samples - 1; ++ii) {
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWellGeneric.cpp:342]: Are we sure that we want to remove the line-intersection algorithms related to solve the THP controlled wells? @totto82 @atgeirr

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 2]
```diff
@@ -2320,10 +2320,43 @@ namespace Opm
...
+            double pressure_cell = this->getPerfCellPressure(fs).value();
+            max_pressure = std::max(max_pressure, pressure_cell);
+        }
+        auto bhpAtLimit = this->StandardWellGeneric<Scalar>::computeBhpAtThpLimitProdWithAlq(frates,
                                                                                   summary_state,
                                                                                   deferred_logger,
+                                                                                  max_pressure,
                                                                                   alq_value);
+        auto v = frates(*bhpAtLimit);
+        if(bhpAtLimit && std::all_of(v.cbegin(), v.cend(), [](double i){ return i <= 0; }))
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWell_impl.hpp:2337]: Should the extra condition, ` std::all_of(v.cbegin(), v.cend(), [](double i){ return i <= 0; })` also involved in the MSW implementation?

Comment (totto82) [reply opm/simulators/wells/StandardWell_impl.hpp:2337]: I think so, but I didn't what to change anything related to MSW in the current PR.

File: opm/simulators/wells/StandardWellGeneric.cpp (hunk 1 above)
Comment (totto82) [reply opm/simulators/wells/StandardWellGeneric.cpp:342]: I am not sure. I did some testing where I added it for both MSW and STW as a last try if everything else failed, but it didn't do much difference.

**Comment by GitPaean** [comment]
I did not do much testing with this PR. But from code review side, I do not have objection to get this in.

**Comment by totto82** [comment]
More testing indicates that this refactoring where the msw code is used for both msw and stw has negative impact on cases with stw. I will investigate. Maybe the msw code needs adjustments before we can use it as a common code for both msw and stw.

**Comment by bska** [comment]
> has negative impact on cases with stw.

What does "negative" mean here?  Worse convergence?  Failure to complete simulation run?  Increased execution time?

**Comment by totto82** [comment]
"negative" = wrong well control for some wells. They switch to THP but should be GRUP. Also wells closing that should not close. It is hard to distinguish what is cause and what is effect here. The code is almost the same so I will try to find out what exactly causes the first difference.

**Comment by totto82** [comment]
>  I will investigate.

Investigation indicates that the reported issue is not directly related to the changes in the code here. I will take a final look at the test failures and start the merging process.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this update_data please

**Comment by totto82** [comment]
jenkins build this opm-tests=740 please
